### PR TITLE
Feature/previous next values

### PR DESCRIPTION
**Pull request by bearrito**
This will address - https://github.com/RoaringBitmap/roaring/issues/380

There a few edge cases I'm not sure about. Namely ones where the the next or previous values would lie completely outside.

### CODE REVIEW

File: bitmapcontainer.go [hunk 1]
```diff
@@ -1234,3 +1234,60 @@ func (bc *bitmapContainer) addOffset(x uint16) (container, container) {
...
+			trailing := countTrailingOnes(bc.bitmap[x])
+			return int(x*64) + trailing
+		}
+
+	}
+	return -1
+}
+
+func (bc *bitmapContainer) previousValue(target uint16) int {
+	return -1
```
Comment (Oppen) [inline bitmapcontainer.go:1292]: I guess this one is incomplete?

**Review by Oppen — CHANGES_REQUESTED** [review]
I think it looks good for the most part. I have a few suggestions and I'd like for the docs to be more clear about the meaning. I had to do some guess work to understand what the functionality achieves, ideally users shouldn't have to do that.

File: setutil_test.go [hunk 2]
```diff
@@ -136,3 +137,94 @@ func TestSetUtilBinarySearch(t *testing.T) {
...
+			assert.Equal(t, testCase.index, result.index)
+			assert.Equal(t, testCase.isExactMatch, result.exactMatch)
+		})
+	}
+}
+
+func TestBinarySearchPastWithBounds(t *testing.T) {
+	type searchTest struct {
+		name         string
+		constructor  func() []uint16
```
[the remark this answers is not in the document]
Comment (bearrito) [reply setutil_test.go:189]: will do.

File: setutil_test.go [hunk 3]
```diff
@@ -136,3 +137,94 @@ func TestSetUtilBinarySearch(t *testing.T) {
 		}
 	}
 }
+
+func TestBinarySearchUntil(t *testing.T) {
+	type searchTest struct {
+		name         string
+		constructor  func() []uint16
```
[the remark this answers is not in the document]
Comment (bearrito) [reply setutil_test.go:144]: ditto.

File: bitmapcontainer.go (hunk 1 above)
Comment (bearrito) [reply bitmapcontainer.go:1292]: yep, everything will be complete when I move out of draft.

File: setutil.go [hunk 4]
```diff
@@ -548,3 +545,121 @@ func binarySearch(array []uint16, ikey uint16) int {
...
+		} else {
+			if middleIndex < maxIndex && target < array[middleIndex+1] {
+				return searchResult{middleValue, middleIndex, false}
+			}
+			lowIndex = middleIndex + 1
+		}
+
+	}
+
+	return searchResult{uint16(closestIndex), closestIndex, false}
```
[the remark this answers is not in the document]
Comment (bearrito) [reply setutil.go:620]: Updated. I should have had test cases covering this.

**Comment by lemire** [comment]
running tests

**Review by Oppen — APPROVED** [review]

**Comment by Oppen** [comment]
Running tests.

**Comment by bearrito** [comment]
Looks like these passed. I still need to implement `previousAbsentValue` for array containers. That should be straightforward.

**Comment by lemire** [comment]
Running tests.

**Comment by bearrito** [comment]
I'll look at the failures. They are on my end.

**Comment by bearrito** [comment]
Tests pass. I don't think I can think of how to apply any more test pressure.

I did have a question about the overall design. I more or less referred to the Java implementation. I had a question about what was done there and how users use this feature. Why do we not for a given container, especially for absent value return values that are in the keyspace of a given key. 

Example: For a run container with HOB X, we can calculate the range of of values that could be under X, call that the image of X. Today if the query/target is less than `min(Union(ranges))` we return `-1` meaning out of bound. Why would we not `min(Union(ranges))-1` if that value was in the image of X.

**Comment by lemire** [comment]
> For a run container with HOB X, we can calculate the range of of values that could be under X, call that the image of X. Today if the query/target is less than min(Union(ranges)) we return -1 meaning out of bound. Why would we not min(Union(ranges))-1 if that value was in the image of X.

Can you elaborate?

What code do you propose changing?

**Comment by lemire** [comment]
I have reopened the issue as I think it was closed by mistake.

Let me go back to the Java functionality...

```Java
  /**
   * Returns the first value equal to or larger than the provided value
   * (interpreted as an unsigned integer). If no such
   * bit exists then {@code -1} is returned. It is not necessarily a
   * computationally effective way to iterate through the values.
   *
   * @param  fromValue the lower bound (inclusive)
   * @return the smallest value larger than or equal to the specified value,
   *       or {@code -1} if there is no such value
   */
  long nextValue(int fromValue);
  
  /**
   * Returns the first value less than or equal to the provided value
   * (interpreted as an unsigned integer). If no such
   * bit exists then {@code -1} is returned. It is not an efficient
   * way to iterate through the values backwards.
   *
   * @param  fromValue the upper bound (inclusive)
   * @return the largest value less than or equal to the specified value,
   *       or {@code -1} if there is no such value
   */
  long previousValue(int fromValue);
```

The current PR only addresses the underlying containers but does not yet exposes the functionality to the users.

In Java, we return a -1 (long) when no value is present.

**Comment by bearrito** [comment]
@lemire yeah I closed it on purpose, so it would not accidentally be merged.

I noticed that as well. I'm implementing that functionality.

I had a few other edge-case bugs. What I did for those was to port all the java tests over and confirm that we match the java tests.

**Comment by lemire** [comment]
@bearrito Your work prompted me to do some work in Javaland:

https://github.com/RoaringBitmap/RoaringBitmap/pull/720

... as it was ill-defined in some edge cases.

**Comment by bearrito** [comment]
@lemire 

I still need to implement and expose `PreviousAbsentValue` via Roaring. 
As mentioned I ported the Java line for line over for runs and arrays. I didn't really understand the Java fluent data builder was doing so I couldn't port for bitmaps. I'm happy with the port, that's a good safety net.

File: roaring.go [hunk 5]
```diff
@@ -1878,6 +1878,185 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
 	rb.highlowcontainer.cloneCopyOnWriteContainers()
 }
 
+func (rb *Bitmap) NextValue(target int) int {
```
Comment (lemire) [inline roaring.go:1881]:
```suggestion
func (rb *Bitmap) NextValue(target uint32) int64 {
```

File: roaring.go [hunk 6]
```diff
@@ -1878,6 +1878,185 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
...
+		} else {
+			prevValue = int(combineLoHi32(uint32(responseBit), uint32(containerKey)))
+		}
+		containerIndex--
+	}
+
+	return prevValue
+}
+
+func (rb *Bitmap) NextAbsentValue(target int) int {
```
Comment (lemire) [inline roaring.go:1959]:
```suggestion
func (rb *Bitmap) NextAbsentValue(target uint32) int64 {
```

File: roaring.go [hunk 7]
```diff
@@ -1878,6 +1878,185 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
...
+		} else {
+			nextValue = int(combineLoHi32(uint32(responseBit), uint32(containerKey)))
+		}
+		containerIndex++
+	}
+
+	return nextValue
+}
+
+func (rb *Bitmap) PreviousValue(target int) int {
```
Comment (lemire) [inline roaring.go:1913]:
```suggestion
func (rb *Bitmap) PreviousValue(target uint32) int64 {
```

File: roaring.go (hunk 5 above)
Comment (lemire) [inline roaring.go:1881]: These new 4 functions should be documented, something like...

```
// NextValue returns the next largest value in the bitmap, or -1
// if none is present. This function should not be used inside
// a performance-sensitive loop: prefer iterators if 
// performance is a concern.
```

Comment (bearrito) [reply roaring.go:1881]: done.

Comment (lemire) [reply roaring.go:1881]: @bearrito I don't see a new commit, did you push your changes?

Comment (bearrito) [reply roaring.go]: pushed. mea culpa.

**Comment by bearrito** [comment]
@lemire 
My fuzz tests are failing. I think I have an overflow issue. But they are passing on my machine

```
--- PASS: TestNextAndPreviousValue (0.05s)
    --- PASS: TestNextAndPreviousValue/randomized (0.05s)
```

Is this an architecture issue? I'll look at it more.

File: roaring.go [hunk 8]
```diff
@@ -1878,6 +1878,201 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
...
+	originalKey := highbits(target)
+	query := lowbits(target)
+	nextValue := -1
+	containerIndex := rb.highlowcontainer.advanceUntil(originalKey, -1)
+	for containerIndex < rb.highlowcontainer.size() && nextValue == -1 {
+		containerKey := rb.highlowcontainer.getKeyAtIndex(containerIndex)
+		container := rb.highlowcontainer.getContainer(containerKey)
+		// if containerKey > orginalKey then we are past the container which mapped to the orignal key
+		// in that case we can just return the minimum from that container
+		var responseBit int
```
Comment (lemire) [inline roaring.go:1895]:
```suggestion
		var responseBit int64
```

File: roaring.go [hunk 9]
```diff
@@ -1878,6 +1878,201 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
...
+		if containerKey > originalKey {
+			bit, err := container.safeMinimum()
+			if err == nil {
+				responseBit = -1
+			}
+			responseBit = int(bit)
```
Comment (lemire) [inline roaring.go:1901]:
```suggestion
			responseBit = int64(bit)
```

File: roaring.go [hunk 10]
```diff
@@ -1878,6 +1878,201 @@ func (rb *Bitmap) CloneCopyOnWriteContainers() {
...
+		} else {
+			responseBit = container.nextValue(query)
+		}
+
+		if responseBit == -1 {
+			nextValue = -1
+		} else {
+			nextValue = int(combineLoHi32(uint32(responseBit), uint32(containerKey)))
```
Comment (lemire) [inline roaring.go:1909]:
```suggestion
			nextValue = int64(combineLoHi32(uint32(responseBit), uint32(containerKey)))
```

**Comment by lemire** [comment]
@bearrito I have marked some suggested changes. The problem is that there is a hidden assumption that `int` spans 64 bits, but it does not. It is either 32 bits or 64 bits. So you should only use `int` when it is ok if `int == int32`.

Generally, it is _safer_ to use `int64` than `int`.

**Comment by lemire** [comment]
> Is this an architecture issue? I'll look at it more.

Yes. It is. On a 32-bit platform they would fail.

Try setting GOARCH to 386 or other 32-bit system.

**Comment by bearrito** [comment]
Thanks. I'm generally poor at low-level bit-thinking/machine sympathy (partly why I wanted to contribute to this project)
I'll correct it.

**Comment by lemire** [comment]
> Thanks. I'm generally poor at low-level bit-thinking/machine sympathy (partly why I wanted to contribute to this project)
I'll correct it.

It is not at all trivial that `int` has a different size depending on the system. :-)

**Comment by bearrito** [comment]
Updated. Tested locally by toggling `GOARCH`

**Comment by lemire** [comment]
Looks good to me. If you recommend merging, we will do so.

**Comment by bearrito** [comment]
I'm satisfied, please merge.
